controlplane/kubeadm/controllers/upgrade.go

```diff
@@ -105,9 +105,17 @@ func (r *KubeadmControlPlaneReconciler) upgradeControlPlane(
 		return ctrl.Result{}, err
 	}
 
-	if status.Nodes <= *kcp.Spec.Replicas {
-		// scaleUp ensures that we don't continue scaling up while waiting for Machines to have NodeRefs
-		return r.scaleUpControlPlane(ctx, cluster, kcp, controlPlane)
+	switch kcp.Spec.RolloutStrategy.Type {
+	case controlplanev1.RollingUpdateStrategyType:
+		// RolloutStrategy is currently defaulted and validated to be RollingUpdate
+		// We can ignore MaxUnavailable because we are enforcing health checks before we get here.
+		maxNodes := *kcp.Spec.Replicas + int32(kcp.Spec.RolloutStrategy.RollingUpdate.MaxSurge.IntValue())
+		if status.Nodes < maxNodes {
```

Comment:
For sake of an excess of prudence:
Are we sure that status.Nodes reflects that current number of nodes (instead of reporting the number of nodes at the previous reconciliation)

Reply:
From a quick check status.Nodes is not the resource status.
However the method which is populating this field is using the cached client, which is not ideal.

I suggest this should use instead the list of machines in the `controlPlane` object, which is populated from the uncached client